lib/wallet/txdb.js

```diff
@@ -512,6 +535,8 @@ class TXDB {
           // possible to compare the on-chain
           // state vs. the mempool state.
           credit.spent = true;
+          // remove coins from 'u'
+          await this.removeSpentCredit(b, credit, path);
```

Comment:
this should also be at line 548?

Reply:
Well, I don't think that it should be at line 548. We have already removed the coin while tx was unconfirmed. Once the tx gets confirmed, the coin has been already removed.

Reply:
Think in terms of asynchronous behavior, it is possible that the `if part` of this condition does not get called.
Example: You get the block containing a transaction before receiving the transaction from the mempool

Reply:
I agree, we can include this line inside `spendCredit()` method.
but should we? because we won't be able to include `saveUnspentCredit()` inside `saveCredit()`
I think it would be better to call it here instead of including it in `spendCredit()`

Reply:
I agree, we can include this line inside `spendCredit()` method.
but should we? because we won't be able to include `saveUnspentCredit()` inside `saveCredit()`
I think it would be better to call it here instead of including it in `spendCredit()`

Reply:
`spendCredit()` should call `removeUnspentCoin()`. Whenever we encounter a TX that consumes one of our UTXO, whether it is in the mempool or in a block, we should remove the unspent coin from `u`.

`saveCredit()` should call `saveUnspentCoin()`. Whenever we encounter a TX that creates a new UTXO for our wallet, we should save it in `u`. When we get a coin in the mempool, we will save a `Coin` with `height: -1` to `u`. And then when we see the same coin in a confirmed TX in a block, `saveUnspentCoin()` will over-write the existing record with a new `Coin` that has `height: <whatever>`. I think this should also be good everywhere `saveCredit()` is called, including `disconnect()` and `erase()`.

What is your doubt?

Reply:
I think it will cause a problem [here](https://github.com/theanmolsharma/bcoin/blob/3413f556c7a89a376bfa6e90c2632249f77aca35/lib/wallet/txdb.js#L539-L540)

here we need to remove the coin from `u` because it is used as an input in a tx but if we use `saveUnspentCoin()` in `saveCredit()` then the coin will get resaved into `u`.

The problem arises because `saveCredit()` is used to save both spent and unspent Credits.
Maybe we can call `saveUnspentCoin()` or `removeUnspentCoin()` from inside of `saveCredit()` based on the value of `spent` flag.